Use SerializedImageBuffer to pass ImageBuffers between workers and the main thread.

mattwoodrow opened the pull request:
#### 2a8263edbfe035e846cb103a41d14a4df3a8d357
<pre>
Use SerializedImageBuffer to pass ImageBuffers between workers and the main thread.
<a href="https://bugs.webkit.org/show_bug.cgi?id=244830">https://bugs.webkit.org/show_bug.cgi?id=244830</a>
&lt;rdar://99857231&gt;

Reviewed by Kimmo Kinnunen.

Create SerializeImageBuffer to transfer ImageBuffers between threads.

Problem: RemoteImageBufferProxy objects cannot be used from any thread other than the one they were
created one. The current code makes a clone into a non-remote ImageBuffer at the locations where
we transfer to/from a worker thread.

The JS Web API here is that when you postMessage an ImageBitmap to a different thread, the existing
object becomes invalid. Thus it is impossible to have concurrent references to an image from multiple
threads.

One option would be to add a way to re-initialize an ImageBuffer object when it is received on another thread.
That&apos;s difficult since it&apos;s hard to ensure that all members and nested objects are handled correctly.

A variant of that would be to re-create the ImageBuffer when it&apos;s received on the other thread. That&apos;s maybe
ok, but still weird that we have a pointer to an object that we&apos;re not allowed to touch. We&apos;d need to post it
back to it&apos;s original thread when we&apos;re done, what if that thread no longer exists?

This takes a slightly different approach of creating a serialization object, and making ImageBuffer non-threadsafe refcounted.
For RemoteImageBufferProxy, we only serialize the identifiers for the remote object, and (via runtime asserts) ensure that the old
object will be consumed in the process.

This is definitely move heavyweight, but I like that it keeps the per-thread instances where they belong. It doesn&apos;t avoid the
complexity of needing to post between threads for failure cases etc.

* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::sinkIntoImageBuffer):
* Source/WebCore/page/Chrome.h:
* Source/WebCore/page/ChromeClient.cpp:
(WebCore::ChromeClient::sinkIntoImageBuffer):
* Source/WebCore/page/ChromeClient.h:
* Source/WebCore/platform/GraphicsClient.h:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::SerializedImageBuffer::sinkIntoImageBuffer):
(WebCore::DefaultSerializedImageBuffer::DefaultSerializedImageBuffer):
(WebCore::ImageBuffer::sinkIntoSerializedImageBuffer):
(WebCore::ImageBuffer::takeBackend):
* Source/WebCore/platform/graphics/ImageBuffer.h:
(WebCore::ImageBuffer::backendInfo):
(WebCore::SerializedImageBuffer::isRemoteSerializedImageBufferProxy const):
(WebCore::SerializedImageBuffer::isSerializedRemoteImageBuffer const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::disconnect):
(WebKit::RemoteDisplayListRecorderProxy::connect):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::RemoteImageBufferProxy):
(WebKit::RemoteImageBufferProxy::sinkIntoSerializedImageBuffer):
(WebKit::RemoteSerializedImageBufferProxy::sinkIntoImageBuffer):
(WebKit::RemoteSerializedImageBufferProxy::~RemoteSerializedImageBufferProxy):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
(WebKit::RemoteImageBufferProxy::create):
(WebKit::RemoteSerializedImageBufferProxy::renderingResourceIdentifier):
(WebKit::RemoteSerializedImageBufferProxy::renderingBackendIdentifier const):
(WebKit::RemoteSerializedImageBufferProxy::RemoteSerializedImageBufferProxy):
(isType):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::~RemoteRenderingBackendProxy):
(WebKit::RemoteRenderingBackendProxy::createRemoteImageBuffer):
(WebKit::RemoteRenderingBackendProxy::transferImageBuffer):
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::clear):
(WebKit::RemoteResourceCacheProxy::releaseImageBuffer):
(WebKit::RemoteResourceCacheProxy::forgetImageBuffer):
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::sinkIntoImageBuffer):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebWorkerClient.cpp:
(WebKit::WebWorkerClient::sinkIntoImageBuffer):
* Source/WebKit/WebProcess/WebCoreSupport/WebWorkerClient.h:

Use ThreadSafeObjectHeap to transfer RemoteImageBuffers between WorkQueues

We&apos;re currently dispatching messages between WorkQueue, which has indeterminate order WRT IPC messages.
Using ThreadSafeObjectHeap should let us store and retrieve the references directly, without waiting in queues.

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::releaseRenderingResource):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
(WebKit::GPUConnectionToWebProcess::RemoteImageBufferData::RemoteImageBufferData):
(WebKit::GPUConnectionToWebProcess::serializedImageHeap):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::moveToSerializedBuffer):
(WebKit::RemoteRenderingBackend::moveToImageBuffer):
(WebKit::RemoteRenderingBackend::transferImageBuffer): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::sinkIntoSerializedImageBuffer):
(WebKit::RemoteSerializedImageBufferProxy::RemoteSerializedImageBufferProxy):
(WebKit::RemoteSerializedImageBufferProxy::sinkIntoImageBuffer):
(WebKit::RemoteSerializedImageBufferProxy::~RemoteSerializedImageBufferProxy):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
(WebKit::RemoteSerializedImageBufferProxy::renderingBackendIdentifier const): Deleted.
(WebKit::RemoteSerializedImageBufferProxy::RemoteSerializedImageBufferProxy): Deleted.
(WebKit::RemoteSerializedImageBufferProxy::sinkIntoImageBufferCompleted): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::moveToSerializedBuffer):
(WebKit::RemoteRenderingBackendProxy::moveToImageBuffer):
(WebKit::RemoteRenderingBackendProxy::transferImageBuffer): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteSerializedImageBufferIdentifier.h: Added.

Use SerializedImageBuffer to pass ImageBuffers between workers and the main thread.

* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneDeserializer::readTransferredImageBitmap):
(WebCore::CloneDeserializer::readOffscreenCanvas):
(WebCore::SerializedScriptValue::create):
* Source/WebCore/html/ImageBitmap.cpp:
(WebCore::ImageBitmap::detachBitmaps):
* Source/WebCore/html/ImageBitmapBacking.cpp:
(WebCore::ImageBitmapBacking::disconnect):
(WebCore::ImageBitmapBacking::connect):
* Source/WebCore/html/ImageBitmapBacking.h:
* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::DetachedOffscreenCanvas::DetachedOffscreenCanvas):
(WebCore::DetachedOffscreenCanvas::takeImageBuffer):
(WebCore::OffscreenCanvas::create):
(WebCore::OffscreenCanvas::pushBufferToPlaceholder):
(WebCore::OffscreenCanvas::commitToPlaceholderCanvas):
(WebCore::OffscreenCanvas::takeImageBuffer const):
* Source/WebCore/html/OffscreenCanvas.h:

Canonical link: <a href="https://commits.webkit.org/257979@main">https://commits.webkit.org/257979@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4947bf093381d8f7d0b0a1312358b889ce6a51fa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100493 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9643 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33545 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109816 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170109 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104481 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10571 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/240 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92904 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107670 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106271 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/7999 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91250 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34623 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/104012 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/89912 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22639 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/77588 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3376 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24158 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3372 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/240 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9491 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43648 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5467 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5183 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebCore/html/ImageBitmapBacking.cpp:
```diff
@@ -62,4 +66,29 @@ unsigned ImageBitmapBacking::height() const
 
+void ImageBitmapBacking::disconnect()
+{
+    // FIXME: It sucks that we have both the ImageBuffer and the SerializedImageBuffer here,
```
"It sucks ..." doesn't add any value.

"I should ..." looks odd when there are multiple contributors to the project.

Form suggestion (can be others)
`// FIXME <what needs to be fixed> <optional rationale if it's not self-evident>
`

kkinnunen-apple commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -4741,8 +4743,10 @@ ExceptionOr<Ref<SerializedScriptValue>> SerializedScriptValue::create(JSGlobalOb
     Vector<std::unique_ptr<DetachedOffscreenCanvas>> detachedCanvases;
-    for (auto offscreenCanvas : offscreenCanvases)
-        detachedCanvases.append(offscreenCanvas->detach());
+    for (auto offscreenCanvas : offscreenCanvases) {
```
I think this is redundant change and a regression in terms of being more complex than the original?

kkinnunen-apple commented on Source/WebKit/WebProcess/WebCoreSupport/WebWorkerClient.cpp:
```diff
@@ -96,6 +96,18 @@ PlatformDisplayID WebWorkerClient::displayID() const
 
+RefPtr<ImageBuffer> WebWorkerClient::sinkIntoImageBuffer(std::unique_ptr<SerializedImageBuffer> imageBuffer)
+{
+#if ENABLE(GPU_PROCESS)
```
Yess.. the polymorphic double dispatch without the shared base "vocabulary".
I wish I knew the solution to this structure..

kkinnunen-apple commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
```diff
@@ -203,6 +206,17 @@ class GPUConnectionToWebProcess
 
     void lowMemoryHandler(WTF::Critical, WTF::Synchronous);
 
+    struct RemoteImageBufferData : public ThreadSafeRefCounted<RemoteImageBufferData> {
```
I think this is good to disambiguate to mean "serialised image buffers". 
So you have

- RemoteSerializedImageBufferProxy
- RemoteSerializedImageBufferIdentifier, Reference, ...

These should be called:
- RemoteSerializedImageBuffer
- RemoteSerializedImageBufferObjectHeap
- remoteSerializedImageBufferObjectHeap()
- m_remoteSerializdeImageBufferObjectHeap

Since GPUConnectionToWebProcess is a massive class and C++ inner classes are cumbersome, you could just introduce the .h, .cpp files for the concepts or alternatively introduce forward reference to RemoteSerializedImageBuffer and then implement it in the GPUConnectionToWebProcess.

kkinnunen-apple commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
```diff
@@ -203,6 +206,17 @@ class GPUConnectionToWebProcess
+        std::unique_ptr<WebCore::ImageBufferBackend> m_backend;
+        WebCore::ImageBufferBackend::Info m_info;
+    };
+    ThreadSafeObjectHeap<RemoteSerializedImageBufferIdentifier, RefPtr<RemoteImageBufferData>>& serializedImageHeap() { return m_imageHeap; }
```
and this could be called `remoteSerializedImageBufferObjectHeap()`

kkinnunen-apple commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.messages.in:
```diff
@@ -25,6 +25,7 @@
 messages -> GPUConnectionToWebProcess WantsDispatchMessage {
     void CreateRenderingBackend(struct WebKit::RemoteRenderingBackendCreationParameters creationParameters, IPC::StreamServerConnection::Handle connectionHandle) AllowedWhenWaitingForSyncReply
     void ReleaseRenderingBackend(WebKit::RenderingBackendIdentifier renderingBackendIdentifier) AllowedWhenWaitingForSyncReply
+    void ReleaseRenderingResource(WebKit::RemoteSerializedImageBufferWriteReference reference)
```
I think this should be called ReleaseSerializedImageBuffer

kkinnunen-apple reviewed (approved):
Please squash and commit as one.

mattwoodrow committed:
Use SerializedImageBuffer to pass ImageBuffers between workers and the main thread.
https://bugs.webkit.org/show_bug.cgi?id=244830
<rdar://99857231>

Reviewed by Kimmo Kinnunen.

Create SerializeImageBuffer to transfer ImageBuffers between threads.

Problem: RemoteImageBufferProxy objects cannot be used from any thread other than the one they were
created one. The current code makes a clone into a non-remote ImageBuffer at the locations where
we transfer to/from a worker thread.

The JS Web API here is that when you postMessage an ImageBitmap to a different thread, the existing
object becomes invalid. Thus it is impossible to have concurrent references to an image from multiple
threads.

One option would be to add a way to re-initialize an ImageBuffer object when it is received on another thread.
That's difficult since it's hard to ensure that all members and nested objects are handled correctly.

A variant of that would be to re-create the ImageBuffer when it's received on the other thread. That's maybe
ok, but still weird that we have a pointer to an object that we're not allowed to touch. We'd need to post it
back to it's original thread when we're done, what if that thread no longer exists?

This takes a slightly different approach of creating a serialization object, and making ImageBuffer non-threadsafe refcounted.
For RemoteImageBufferProxy, we only serialize the identifiers for the remote object, and (via runtime asserts) ensure that the old
object will be consumed in the process.

This is definitely move heavyweight, but I like that it keeps the per-thread instances where they belong. It doesn't avoid the
complexity of needing to post between threads for failure cases etc.

* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::sinkIntoImageBuffer):
* Source/WebCore/page/Chrome.h:
* Source/WebCore/page/ChromeClient.cpp:
(WebCore::ChromeClient::sinkIntoImageBuffer):
* Source/WebCore/page/ChromeClient.h:
* Source/WebCore/platform/GraphicsClient.h:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::SerializedImageBuffer::sinkIntoImageBuffer):
(WebCore::DefaultSerializedImageBuffer::DefaultSerializedImageBuffer):
(WebCore::ImageBuffer::sinkIntoSerializedImageBuffer):
(WebCore::ImageBuffer::takeBackend):
* Source/WebCore/platform/graphics/ImageBuffer.h:
(WebCore::ImageBuffer::backendInfo):
(WebCore::SerializedImageBuffer::isRemoteSerializedImageBufferProxy const):
(WebCore::SerializedImageBuffer::isSerializedRemoteImageBuffer const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::disconnect):
(WebKit::RemoteDisplayListRecorderProxy::connect):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::RemoteImageBufferProxy):
(WebKit::RemoteImageBufferProxy::sinkIntoSerializedImageBuffer):
(WebKit::RemoteSerializedImageBufferProxy::sinkIntoImageBuffer):
(WebKit::RemoteSerializedImageBufferProxy::~RemoteSerializedImageBufferProxy):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
(WebKit::RemoteImageBufferProxy::create):
(WebKit::RemoteSerializedImageBufferProxy::renderingResourceIdentifier):
(WebKit::RemoteSerializedImageBufferProxy::renderingBackendIdentifier const):
(WebKit::RemoteSerializedImageBufferProxy::RemoteSerializedImageBufferProxy):
(isType):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::~RemoteRenderingBackendProxy):
(WebKit::RemoteRenderingBackendProxy::createRemoteImageBuffer):
(WebKit::RemoteRenderingBackendProxy::transferImageBuffer):
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::clear):
(WebKit::RemoteResourceCacheProxy::releaseImageBuffer):
(WebKit::RemoteResourceCacheProxy::forgetImageBuffer):
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::sinkIntoImageBuffer):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebWorkerClient.cpp:
(WebKit::WebWorkerClient::sinkIntoImageBuffer):
* Source/WebKit/WebProcess/WebCoreSupport/WebWorkerClient.h:

Use ThreadSafeObjectHeap to transfer RemoteImageBuffers between WorkQueues

We're currently dispatching messages between WorkQueue, which has indeterminate order WRT IPC messages.
Using ThreadSafeObjectHeap should let us store and retrieve the references directly, without waiting in queues.

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::releaseRenderingResource):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
(WebKit::GPUConnectionToWebProcess::RemoteImageBufferData::RemoteImageBufferData):
(WebKit::GPUConnectionToWebProcess::serializedImageHeap):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::moveToSerializedBuffer):
(WebKit::RemoteRenderingBackend::moveToImageBuffer):
(WebKit::RemoteRenderingBackend::transferImageBuffer): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::sinkIntoSerializedImageBuffer):
(WebKit::RemoteSerializedImageBufferProxy::RemoteSerializedImageBufferProxy):
(WebKit::RemoteSerializedImageBufferProxy::sinkIntoImageBuffer):
(WebKit::RemoteSerializedImageBufferProxy::~RemoteSerializedImageBufferProxy):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
(WebKit::RemoteSerializedImageBufferProxy::renderingBackendIdentifier const): Deleted.
(WebKit::RemoteSerializedImageBufferProxy::RemoteSerializedImageBufferProxy): Deleted.
(WebKit::RemoteSerializedImageBufferProxy::sinkIntoImageBufferCompleted): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::moveToSerializedBuffer):
(WebKit::RemoteRenderingBackendProxy::moveToImageBuffer):
(WebKit::RemoteRenderingBackendProxy::transferImageBuffer): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteSerializedImageBufferIdentifier.h: Added.

Use SerializedImageBuffer to pass ImageBuffers between workers and the main thread.

* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneDeserializer::readTransferredImageBitmap):
(WebCore::CloneDeserializer::readOffscreenCanvas):
(WebCore::SerializedScriptValue::create):
* Source/WebCore/html/ImageBitmap.cpp:
(WebCore::ImageBitmap::detachBitmaps):
* Source/WebCore/html/ImageBitmapBacking.cpp:
(WebCore::ImageBitmapBacking::disconnect):
(WebCore::ImageBitmapBacking::connect):
* Source/WebCore/html/ImageBitmapBacking.h:
* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::DetachedOffscreenCanvas::DetachedOffscreenCanvas):
(WebCore::DetachedOffscreenCanvas::takeImageBuffer):
(WebCore::OffscreenCanvas::create):
(WebCore::OffscreenCanvas::pushBufferToPlaceholder):
(WebCore::OffscreenCanvas::commitToPlaceholderCanvas):
(WebCore::OffscreenCanvas::takeImageBuffer const):
* Source/WebCore/html/OffscreenCanvas.h:

Canonical link: https://commits.webkit.org/257979@main
